runtime/src/iree/hal/drivers/cuda/cuda_device.h

```diff
@@ -23,6 +23,11 @@ iree_status_t iree_hal_cuda_device_create(
     iree_hal_cuda_dynamic_symbols_t* syms, CUdevice device,
     iree_allocator_t host_allocator, iree_hal_device_t** out_device);
 
+// Returns a CUDA context bound to the given `base_device` if it is a HAL CUDA
+// device. Returns error if `base_device` is not a HAL CUDA device.
+iree_status_t iree_hal_cuda_device_get_context(iree_hal_device_t* base_device,
+                                               CUcontext* out_context);
```

Comment:
I'm not fully sure how you'll want to set up the custom module to interop with this, but you _could_ invert this by creating the module _first_, then sharing the module's `CUcontext` with the CUDA driver.

The Vulkan HAL has an `api.h` file with high level functions for sharing context, such as this one: https://github.com/openxla/iree/blob/5593180c12217fbfa3fea49fc98ed51bf36dbf21/runtime/src/iree/hal/drivers/vulkan/api.h#L266-L277

The way you have the module set up right now expects that there is only a single CUDA device per application session and that device outlives the module. There could be multiple devices and their lifetimes could vary.

cc @benvanik

Reply:
The idea is that it should work out of the box for N modules defined in M separate projects requiring CUDA device, with `iree-module-runner` lazy loading modules (similar to `iree-compiler` loading compiler plugins), it's going to be tricky to agree between all of the modules how to create context (e.g. how to configure context flags).

Regarding multiple-devices + modules - this is a limitation of custom module registration API, once it will support multiple devices, my custom module will just find one or more compatible devices, and keep `CUContext` for them.